src/main/java/com/example/bookshop/service/impl/BookServiceImpl.java

```diff
@@ -82,21 +82,22 @@ public List<BookDto> search(BookSearchParameters bookSearchParameters, Pageable
                 .toList();
     }
 
-    private Book saveOrUpdateCategories(Book book, CreateBookRequestDto requestDto) {
-        Set<Category> categories = requestDto.categoriesIds().stream()
-                .map(categoryRepository::findById)
-                .map(o -> o.orElseThrow(() -> new EntityNotFoundException("Can't"
-                        + " find category.")))
-                .collect(Collectors.toSet());
-        book.setCategories(categories);
-        return book;
-    }
-
     @Override
-    public List<BookDto> getAllBooksByCategory(Long id) {
-        return bookRepository.findAllByCategoriesId(id)
-                .stream()
+    public List<BookDto> getAllBooksByCategory(Long id, Pageable pageable) {
+        return bookRepository.findAllByCategoriesId(id, pageable).stream()
                 .map(bookMapper::toDto)
                 .toList();
     }
+
+    private Book saveOrUpdateCategories(Book book, CreateBookRequestDto requestDto) {
+        if (requestDto.categoriesIds() != null) {

```

Comment:
you can refactor this method and avoid this construction. I recommend you use the annotation above the field categoryIds in Dto